specs/crux/beacon-chain.md

```diff
@@ -0,0 +1,247 @@
+    - [`Validator`](#validator)
+    - [`BeaconBlockBody`](#beaconblockbody)
+  - [New containers](#new-containers)
+    - [`DelegationMessage`](#delegationmessage)
+    - [`Delegation`](#delegation)
+- [Beacon chain state transition function](#beacon-chain-state-transition-function)
+  - [Block processing](#block-processing)
+    - [Operations](#operations)
+    - [Delegation](#delegation)
+      - [`get_validator_from_delegation`](#get_validator_from_delegation)
+      - [`process_delegation`](#process_delegation)
+  - [Epoch processing](#epoch-processing)
+    - [`process_delegation_transfers`](#process_delegation_transfers)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Introduction
+
+This upgrade adds validator delegation. A validator may choose to delegate all its earnings above `MAX_EFFECTIVE_BALANCE` to an existing or a new validator.
+
+## Constants
+
+### Domain types
+
+| Name  | Value |
+| - | - |
+| `DOMAIN_DELEGATION` | `DomainType('0x0A000000')` |
+
+## Preset
+
+### Max operations per block
+
+| Name | Value |
+| - | - |
+| `MAX_DELEGATIONS` | `2**4` (=16) |
+
+### Time parameters
+| Name | Value | Unit | Duration | 
+| - | - | :-: | :-: |
+| `MAX_DELEGATION_INCLUSION_DELAY` | `uint64(2**7)` (=128) | slots | 25.6 minutes | 
+
+## Configuration
+
+### Gwei values
+
+| Name | Value |
+| - | - |
+| `DELEGATION_TRANSACTION_COST` | `Gwei(10**6)` (= 1,000,000) TBD |
+
+
+## Containers
+
+### Extended containers
+
+#### `Validator`
+The `Validator` container gains a new field `delegate` containing the validator index of the delegate validator. This field is initialized at the fork with the same validator index of the current validator. (N.B. implementers may chose to keep a separate map instead)
+
+```python
+class Validator(Container):
+    pubkey: BLSPubkey
+    delegate: ValidatorIndex  # [New in Crux]
+    withdrawal_credentials: Bytes32
+    effective_balance: Gwei
+    slashed: boolean
+    # Status epochs
+    activation_eligibility_epoch: Epoch
+    activation_epoch: Epoch
+    exit_epoch: Epoch
+    withdrawable_epoch: Epoch
+```
+
+#### `BeaconBlockBody`
+
+```python
+class BeaconBlockBody(Container):
+    randao_reveal: BLSSignature
+    eth1_data: Eth1Data
+    graffiti: Bytes32
+    # Operations
+    proposer_slashings: List[ProposerSlashing, MAX_PROPOSER_SLASHINGS]
+    attester_slashings: List[AttesterSlashing, MAX_ATTESTER_SLASHINGS]
+    attestations: List[Attestation, MAX_ATTESTATIONS]
+    deposits: List[Deposit, MAX_DEPOSITS]
+    voluntary_exits: List[SignedVoluntaryExit, MAX_VOLUNTARY_EXITS]
+    delegations: List[Delegation, MAX_DELEGATIONS]  # [New in Crux]
+    sync_aggregate: SyncAggregate
+    # Execution
+    execution_payload: ExecutionPayload
+```
+
+### New containers
+
+#### `DelegationMessage`
+
+```python
+class DelegationMessage(Container):
+    delegating_index: ValidatorIndex
+    delegating_pubkey: BLSPubkey
```

Comment:
Why is delegating_pubkey necessary?

Reply:
It's the withrawal public key that signs the message